Add functional collective all_to_all_single and support it in Inductor

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110195

Copy of https://github.com/pytorch/pytorch/pull/106655 from yf225
rebased on top of item() support changes

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

ezyang committed:
Add functional collective all_to_all_single and support it in Inductor

Copy of https://github.com/pytorch/pytorch/pull/106655 from yf225
rebased on top of item() support changes

[ghstack-poisoned]

ezyang commented:
cc @yf225

ezyang commented:
Hmm I think I didn't do the merge update completely correctly

ezyang committed:
Update on "Add functional collective all_to_all_single and support it in Inductor"


Copy of https://github.com/pytorch/pytorch/pull/106655 from yf225
rebased on top of item() support changes

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang committed:
Update on "Add functional collective all_to_all_single and support it in Inductor"


Copy of https://github.com/pytorch/pytorch/pull/106655 from yf225
rebased on top of item() support changes

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang committed:
Update on "Add functional collective all_to_all_single and support it in Inductor"


Copy of https://github.com/pytorch/pytorch/pull/106655 from yf225
rebased on top of item() support changes

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang committed:
Update on "Add functional collective all_to_all_single and support it in Inductor"


Copy of https://github.com/pytorch/pytorch/pull/106655 from yf225
rebased on top of item() support changes

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang committed:
Update on "Add functional collective all_to_all_single and support it in Inductor"


Copy of https://github.com/pytorch/pytorch/pull/106655 from yf225
rebased on top of item() support changes

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang committed:
Update on "Add functional collective all_to_all_single and support it in Inductor"


Copy of https://github.com/pytorch/pytorch/pull/106655 from yf225
rebased on top of item() support changes

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

yf225 commented on test/distributed/test_inductor_collectives.py:
```diff
@@ -261,6 +267,170 @@ def compile(func, example_inputs):
+            compiled_fn = torch.compile(example, fullgraph=True, dynamic=True)
+            code = run_and_get_triton_code(compiled_fn, *inputs, **trs)
+            FileCheck() \
+                .check("fun_col_impl._all_to_all_single") \
```
should we assert that generated code has the unbacked symints at the right places?

ezyang replied:
it was a pain manually updating the FileChecks when I wobbled the code lol. Can we come up with a less brittle way to test this

yf225 replied:
perhaps a regex matching the prefix `i` but not the actual variable number? I can add it after this PR

ezyang committed:
Update on "Add functional collective all_to_all_single and support it in Inductor"


Copy of https://github.com/pytorch/pytorch/pull/106655 from yf225
rebased on top of item() support changes

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang commented on torch/_inductor/ir.py:
```diff
@@ -4896,6 +4930,7 @@ def get_strides_of_lstm_output(output_shape, batch_first):
                 zip(output_sizes, output_strides)
             )
         ]
+        packed.outputs = output_ir
```
I guess I don't need this anymore

ezyang commented on torch/distributed/_functional_collectives.py:
```diff
@@ -509,6 +541,17 @@ def mk_out_tensor(input):
 
     return [mk_out_tensor(t) for t in inputs]
 
+# TODO: think, this isn't actually dynamic?!
```
I was surprised here

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2329,6 +2329,8 @@ def _init(
         # Currently populated from tensors
         self.var_to_val: Dict[sympy.Symbol, sympy.Integer] = {}
+        # Maps symbolic ints to their size hints. Only used for unbacked symints.
+        self.var_to_size_hint: Dict[sympy.Symbol, sympy.Integer] = {}
```
still not convinced by this

Skylion007 commented on torch/_inductor/ir.py:
```diff
@@ -2486,6 +2524,23 @@ def get_unbacked_symbol_uses(self):
         return set()
 
+    def codegen_unbacked_symbol_defs(self, wrapper):
+        symbols_to_define = self.get_unbacked_symbol_defs()
```
Hmm, if it wasn't for the logging here, I would advise rewriting this as set intersection / difference ops.

ezyang commented:
@pytorchbot merge

ezyang commented on torch/_inductor/ir.py:
```diff
@@ -6213,3 +6256,106 @@ def maybe_free_unbacked_symbols(s):
+        return False
+
+    def should_emit_find_or_create_pg(self):
+        return False
```
@yf225 can you explain why you needed to disable register tensor work / find-create pg? This seems surprising to me

ezyang committed:
Update on "Add functional collective all_to_all_single and support it in Inductor"


Copy of https://github.com/pytorch/pytorch/pull/106655 from yf225
rebased on top of item() support changes

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang committed:
Update on "Add functional collective all_to_all_single and support it in Inductor"


Copy of https://github.com/pytorch/pytorch/pull/106655 from yf225
rebased on top of item() support changes

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

ezyang commented:
@yf225 because the kernel is no longer "dynamic", it can be implemented exactly symmetrically to all the other collectives. I've updated the implementation accordingly.

ezyang commented:
@pytorchbot merge
